fix: function overload in MSV

skimaharvey opened the pull request:
# What does this PR introduce ? 

This PR addresses critical interface compatibility issues between the Solidity implementation and the original Vyper vault contracts. The changes ensure 1:1 compatibility with the Vyper implementation while maintaining ERC4626 compliance.

## Changes


### 1. Missing Withdraw/Redeem Function Overloads

Added missing function overloads that were not originally ported from Vyper:

#### Withdraw Functions && Redeem functions: 

- Missing the  4-parameter overloads (documented)

#### MaxWithdraw/MaxRedeem Functions:
- Added 1-parameter and 2-paremeters overloads for both functions with sensible defaults

**Design Decision**: The Vyper implementation includes additional 4-parameter overloads (with just maxLoss parameter). These were intentionally omitted due to Solidity contract size constraints. The 3 and 5 parameter versions provide equivalent functionality while keeping the contract size manageable.

skimaharvey commented on src/core/MultistrategyLockedVault.sol:
```diff
@@ -360,14 +360,62 @@ contract MultistrategyLockedVault is MultistrategyVault, IMultistrategyLockedVau
+        address receiver,
+        address owner
+    ) external override(MultistrategyVault, IMultistrategyVault) returns (uint256) {
+        return withdraw(assets, receiver, owner, 10_000, new address[](0));
```
use 0 instead 10_000

skimaharvey commented on src/core/MultistrategyLockedVault.sol:
```diff
@@ -512,6 +609,12 @@ contract MultistrategyLockedVault is MultistrategyVault, IMultistrategyLockedVau
+    function maxWithdraw(
+        address owner_
+    ) external view override(MultistrategyVault, IMultistrategyVault) returns (uint256) {
+        return maxWithdraw(owner_, MAX_BPS, new address[](0));
```
here it should be maxLoss 0

skimaharvey committed:
refactor: added to natspecs design decision

skimaharvey committed:
refactor: added missing 2 parameters maxWithdraw/maxRedeem

skimaharvey commented on src/core/interfaces/IMultistrategyVault.sol:
```diff
@@ -446,36 +470,39 @@ interface IMultistrategyVault {
+    function set_deposit_limit_module(address depositLimitModule, bool shouldOverride) external;
+    function set_withdraw_limit_module(address withdrawLimitModule) external;
+    function set_minimum_total_idle(uint256 minimumTotalIdle) external;
     function setProfitMaxUnlockTime(uint256 newProfitMaxUnlockTime) external;
```
it is actually camelCased there 

https://github.com/yearn/yearn-vaults-v3/blob/104a2b233bc6d43ba40720d68355b04d2dc31795/contracts/VaultV3.vy#L1488C5-L1488C27

skimaharvey commented on src/core/interfaces/IMultistrategyVault.sol:
```diff
@@ -446,36 +470,39 @@ interface IMultistrategyVault {
     // Management Functions
     function setName(string memory name) external;
     function setSymbol(string memory symbol) external;
```
both are camelCase https://github.com/yearn/yearn-vaults-v3/blob/104a2b233bc6d43ba40720d68355b04d2dc31795/contracts/VaultV3.vy#L1330-L1347

qGolem reviewed (changes requested):
split snake case into its own pr to make it easier to review

skimaharvey committed:
chore: update optimizers runs to 3

skimaharvey committed:
chore: deny warnings

skimaharvey committed:
refactor: optimize MSLV contract size

0xferit reviewed (commented):
deleted

0xferit reviewed (changes requested):
## Findings

### MEDIUM-1: `assess_share_of_unrealised_losses` Breaks Vyper Compatibility

**Files:**
- `src/core/MultistrategyVault.sol` (lines 1926-1933)
- `src/core/interfaces/IMultistrategyVault.sol` (line 406)

**Description:** PR 333 **removes** the 2-param Vyper-compatible wrapper that existed on develop, leaving only the 3-param version:

| Branch | Versions Available |
|--------|-------------------|
| **develop** | ✅ 2-param + 3-param overloads |
| **PR 333** | ❌ Only 3-param (2-param removed) |

**Vyper VaultV3.vy external API (2 params):**
```python
def assess_share_of_unrealised_losses(strategy: address, assets_needed: uint256) -> uint256:
    current_debt: uint256 = self.strategies[strategy].current_debt  # looks up internally
    return self._assess_share_of_unrealised_losses(strategy, current_debt, assets_needed)
```

**develop branch Solidity (before PR 333) - had matching 2-param wrapper:**
```solidity
function assess_share_of_unrealised_losses(address strategy_, uint256 assetsNeeded_) external view returns (uint256) {
    uint256 currentDebt = _strategies[strategy_].currentDebt;  // looks up internally
    return _assess_share_of_unrealised_losses(strategy_, currentDebt, assetsNeeded_);
}
```

**PR 333 removes this wrapper**, leaving only:
```solidity
function assess_share_of_unrealised_losses(address strategy, uint256 currentDebt, uint256 assetsNeeded)
```

**Impact:**
1. **Breaks Vyper ABI compatibility** - contradicts the PR's stated goal
2. **Breaks existing integrations** expecting the 2-param signature

**Risk:** Medium - breaks existing integrations and contradicts the PR's stated goal of Vyper interface compatibility.

---

### MEDIUM-2: Interface Gap for `get_default_queue()`

**Files:**
- `src/core/MultistrategyVault.sol` (line 715)
- `src/core/interfaces/IMultistrategyVault.sol`

**Description:** The function `defaultQueue()` was removed and replaced with `get_default_queue()` which **correctly matches the Vyper function name**. However, `get_default_queue()` is **not declared in the interface**.

**Vyper verification:**
```python
# VaultV3.vy has:
def get_default_queue() -> DynArray[address, MAX_QUEUE]:
```

**Risk:** 
1. Breaking change for integrators using `defaultQueue()` (was non-standard name)
2. Interface consumers cannot call `get_default_queue()` through `IMultistrategyVault`

---

### LOW-1: Utility Functions Removed

**File:** `src/core/MultistrategyLockedVault.sol`

**Description:** Two helper functions removed:
- `getTransferableShares(address user)` 
- `getRageQuitableShares(address user)`

Users must now calculate these values from `custodyInfo(user).lockedShares` manually.

**Risk:** Convenience regression for integrators. NatSpec in the contract explains the manual calculation approach, but no external documentation was added.

---

### LOW-2: NatSpec/Implementation Mismatch for `maxLoss` Default

**File:** `src/core/MultistrategyLockedVault.sol` (lines 363-380)

**Description:** The NatSpec for the 3-parameter `withdraw` function documents `maxLoss = 10000` (100% loss accepted) but the implementation passes `maxLoss = 0` (no loss accepted):

```solidity
/**
 * @notice Withdraws assets with default parameters (maxLoss = 10000, default queue)
 */
function withdraw(uint256 assets, address receiver, address owner) ... {
    ...
    _redeem(msg.sender, receiver, owner, assets, shares, 0, new address[](0));
    //                                               ^-- actual value is 0
}
```

**Analysis:** The code is correct - it follows the same pattern as the parent `MultistrategyVault` where `withdraw` uses `maxLoss=0` and `redeem` uses `maxLoss=10000`. This matches ERC4626 semantics: `withdraw(assets)` specifies exact assets (no slippage tolerance), while `redeem(shares)` burns exact shares (accepts any resulting assets). The NatSpec is simply a typo.

**Risk:** Documentation error only. No fund loss possible - at worst, users experience unexpected reverts and must use the 5-parameter version with explicit `maxLoss`.

skimaharvey committed:
refactor: make assess_share_of_unrealised_losses 2 params

skimaharvey committed:
refactor: add get_default_queue to IMultistrategy

skimaharvey committed:
refactor: update natspecs in MSLV withdraw func

skimaharvey commented:
Thank you for the report.


Med 1 -> good catch and fixed
med 2 -> fixed, added it to the interface 
low 1 -> unfortunately the MSLV does not fit with these 2 functions (they can be implemented in a separate contract)
low 2 -> fixed

skimaharvey committed:
refactor: make RAGE_QUIT_COOLDOWN_CHANGE_DELAY in MSLV

0xferit commented:
Let's merge or close this; it's been months. @qGolem @skimaharvey

skimaharvey commented:
fine to merge for me

qGolem commented:
fix ci and ask for review again

skimaharvey committed:
chore: merge develop and reduce MultistrategyLockedVault below EIP-170 limit
